{% extends 'mainLayout.html' %}
{% block title %}Resources{% endblock %}
{% block content %}
<br />
<br />
<br />
<h3>Microbiome analysis tools</h3>

<table class="table table-hover">
<thead>
<tr>
<th scope="col">Resouces</th>
<th scope="col">Description</th>
<th scope="col">Links</th>
<!--th scope="col">Links</th-->
</tr>
</thead>

<tbody>
<tr>
<td>Tools for Microbiome analysis</td>
<td>Useful R packages  for 16S rRNA gene and shotgun data analysis and visualization</td>
<td>https://microsud.github.io/Tools-Microbiome-Analysis/</td>
</tr>
<tr>
<td>Quantitative Insights Into Microbial Ecology (QIIME)
</td>
<td>QIIME is a bioinformatics pipeline enabling analysis of microbiome DNA metadata. It allows to perform statistical analyses of results and generate high-quality graphs</td>
<td>https://qiime.org/</td>
</tr>
<tr>
<td>dada2 R package</td>
<td>
dada2 is an R package dedicated to the analysis of microbiome data
</td>
<td>http://benjjneb.github.io/dada2/</td>
</tr>
</table>
<br /><br />


<h3>Training Materials</h3>

<table class="table table-hover">
<thead>
<tr>
<th scope="col">Resouces</th>
<th scope="col">Description</th>
<th>Links</th>
</tr>
</thead>
<tbody>


<tr>
<td>16S workflow</td>
<td>H3ABioNet guideline for Bioinformatic and statistical analysis on 16S rDNA data</td>
<td>https://h3abionet.org/tools-and-services/workflows</td>
</tr>
<tr>
<td>
Shotgun Metagenomic analysis tutorial</td>
<td>Shotgun Metagenomic analysis tutorial</td>
<td>http://rstudio-pubs-static.s3.amazonaws.com/425650_32b00adefb29485ba098d3bf9c6b5f82.html#10)_binning_(optional)</td>
</tr>



<tr>
<td>The Complete Guide to the Science of the Microbiome</td>
<td>An introduction to the microbiome science as well as an orientation text about many important/different research aspects</td>
<td>https://endpoints.elysiumhealth.com/microbiome-explainer-e345658db2c</td>
</tr>
<tr>
<td>Microbiome Therapeutics</td>
<td>It contains a useful websites related to the area of microbiome research</td>
<td>http://www.mbiome.com/resources/websites/</td>
</tr>
<tr>
<td>The National Academic Press</td>
<td>Current methods for studying human microbiome project- Human and environmental chemicals interactions characterisation.</td>
<td>https://www.nap.edu/read/24960/chapter/6</td>
</tr>
<tr>
<td>NIH: Teaching the microbiome</td>
<td>Collection of various Internet-based tools and resources for teaching the microbiome</td>
<td>https://www.genome.gov/27552808/teaching-the-microbiome/</td>
</tr>
</table>
<br /><br />

<h3>Microbiome related databases</h3>

<table class="table table-hover">
<thead>
<tr>
<th scope="col">Resouces</th>
<th scope="col">Description</th>
<th scope="col">Links</th>
</tr>
</thead>
<tbody>
<tr>
<td>Human OncoBiome Database</td>
<td>Database of comprehensive curated information on the bacterial species present in the human cancer microbiota</td>
<td>http://www.juit.ac.in/hcmd/home</td>
</tr>
<tr>
<td>MicrobiomeDB</td>
<td>MicrobiomeDB is a data mining platform enabling the exploration of microbiome studies. This platform also provides visualization and analysis tools to investigate microbiome data.</td>
<td>https://microbiomedb.org/mbio/app/</td>
</tr>
<tr>
<td>MDB: Microbiome Database</td>
<td>The microbiome Database gathers microbiome metadata. about the environment (oceanic, soil…) and host-associated microbiome. It also makes available the gene catalogs for host-associated microbiomes.</td>
<td>https://db.cngb.org/microbiome/</td>
</tr>
<tr>
<td>Human Microbiome Project Data Portal</td>
<td>The Human Microbiome Project Data Portal enables the exploration of studies and data associated with the human microbiome. Furthermore, this web interface provides access to the tools used to analyze microbiome.</td>
<td>https://portal.hmpdacc.org/</td>
</tr>
<tr>
<td>Human Oral Microbiome Database (eHOMD)</td>
<td>comprehensive and curated information o­n the bacterial species present in the human aerodigestive tract</td>
<td>http://www.homd.org/</td>
</tr>
<tr>
<td>Disbiome database</td>
<td>Database providing an overview of differential microbial composition in diseases</td>
<td>https://disbiome.ugent.be/home</td>
</tr>
<tr>
<td>MDB: Microbiome Database</td>
<td>This database provides sequencing resource and  accurate metadata of metagenomics samples, as well as gene catalogs.</td>
<td>https://db.cngb.org/microbiome/</td>
</tr>
<tr>
<td>Human Pan-Microbe Communities (HPMC) database</td>
<td>Database of human gastrointestinal microbiota derived from faecal samples.</td>
<td>http://www.hpmcd.org/index.php</td>
</tr>
</table>
<br /><br />



{% comment %}<h4>News and Events</h4>{% endcomment %}


{% comment %}<h5>Upcoming and Current Training</h5>{% endcomment %}
{% comment %}<table class="table table-hover">{% endcomment %}
{% comment %}<thead>{% endcomment %}
{% comment %}<tr>{% endcomment %}
{% comment %}<th scope="col">Resouces</th>{% endcomment %}
{% comment %}<th scope="col">Description</th>{% endcomment %}
{% comment %}</tr>{% endcomment %}
{% comment %}</thead>{% endcomment %}
{% comment %}<tbody>{% endcomment %}
{% comment %}<tr>{% endcomment %}
{% comment %}<td><a href="https://www.h3abionet.org/categories/event/course/16srrna-microbiome-intermediate-bioinformatics-training-20211">16S rRNA Microbiome Intermediate Bioinformatics Training 2021</a></td>{% endcomment %}
{% comment %}<td>The H3ABioNet 16S rRNA Microbiome Intermediate Bioinformatics course will provide training to enable participants to gain the knowledge and skills to perform 16S rRNA microbiome data analyses using a variety of bioinformatics methods and tools.</td>{% endcomment %}
{% comment %}</tr>{% endcomment %}
{% comment %}</tbody>{% endcomment %}
{% comment %}</table>{% endcomment %}

{% comment %}<p>H3ABionet events will added here.</p>{% endcomment %}
{% endblock %}
